components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,402 @@
+                     std::string& field) {
+  CHECK(tokenizer);
+  if (!CheckField(input, msg_token)) {
+    return false;
+  }
+  field = input.substr(msg_token.length());
+  if (field.empty()) {
+    return false;
+  }
+  if (consume_delim && !ConsumeDelim(tokenizer)) {
+    return false;
+  }
+  return true;
+}
+
+bool FillOptionalStringField(StringTokenizer* tokenizer,
+                             const std::string& msg_token,
+                             absl::optional<std::string>& field) {
+  CHECK(tokenizer);
+  std::string field_str;
+  if (!FillStringField(tokenizer, tokenizer->token(), msg_token, false,
+                       field_str)) {
+    return false;
+  }
+  field = field_str;
+  return true;
+}
+
+bool FillOptionalResourcesField(StringTokenizer* tokenizer,
+                                absl::optional<std::vector<GURL>>& field) {
+  CHECK(tokenizer);
+  if (!CheckField(tokenizer->token(), kResourcesToken)) {
+    return false;
+  }
+  std::vector<GURL> urls;
+  while (tokenizer->GetNext()) {
+    if (tokenizer->token_is_delim()) {
+      continue;
+    }
+    if (!CheckField(tokenizer->token(), kResourcesSeperator)) {
+      return false;
+    }
+    auto url_str = tokenizer->token().substr(strlen(kResourcesSeperator));
+    GURL url(url_str);
+    if (!url.is_valid()) {
+      return false;
+    }
+    urls.push_back(url);
+  }
+  if (!urls.size()) {
+    return false;
+  }
+  field = urls;
+  return true;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-2.3
+bool IsUnreservedChar(char c) {
+  if (base::IsAsciiAlphaNumeric(c)) {
+    return true;
+  }
+  switch (c) {
+    case '-':
+    case '.':
+    case '_':
+    case '~':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986#section-2.2
+bool IsSubDelimChar(char c) {
+  switch (c) {
+    case '!':
+    case '$':
+    case '&':
+    case '\'':
+    case '(':
+    case ')':
+    case '*':
+    case '+':
+    case ',':
+    case ';':
+    case '=':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-3.3
+bool IsPChar(char c) {
+  if (IsUnreservedChar(c)) {
+    return true;
+  }
+  if (IsSubDelimChar(c)) {
+    return true;
+  }
+  switch (c) {
+    case ':':
+    case '[':
+    case ']':
+    case '@':
+      return true;
+  }
+  return false;
+}
+
+}  // namespace
+
+mojom::SIWEMessagePtr SIWEMessageParser::Parse(const std::string& message) {
+  state_ = State::kStart;
+  StringTokenizer tokenizer(message, "\n");
+  tokenizer.set_options(StringTokenizer::RETURN_DELIMS);
+  auto result = mojom::SIWEMessage::New();
+  bool seeking_optional_fields = false;
+  while (tokenizer.GetNext() || seeking_optional_fields) {
+    const std::string& token = tokenizer.token();
+    switch (state_) {
+      case State::kStart: {
+        std::string::size_type n = token.find(kStartingToken);
+        if (!n || n == std::string::npos) {
```

Comment:
SIWE message format is not recursive and line based.
I'd suggest that parsing is done like
```
if (!ParseSchemeAndDomain(&message_span)) return false;
if (!ParseAddress(&message_span)) return false;
if (!ParseStatement(&message_span)) return false;
if (!ParseUri(&message_span)) return false;
...
if (!ParseRequestId(&message_span)) return false;
if (!ParseResources(&message_span)) return false;
```
Each call extracts a line from message_span(altering it), checks for prefix and fills data in `this`.

Reply:
Tokenizer here is not recursive, and there are optional fields comes with different `\n` requirement. Tokenized it by '\n` and examine each token would be cleaner.

Reply:
I mean there is a while loop which on every iteration starts parsing "from scratch" with some context from state_. That approach is ok when grammar is recursive, but superfluous for SIWE grammar which is purely sequential.

>  different \n requirement
It's just a weird way to express in grammar that whole message does not need \n at the very end.

Reply:
Each iteration is not parsing from scratch, it is a pre-processed result and the loop just iterate through the each one split by LF.

Reply:
simplify it in baeed212719cd3abd1ec20275916e3e2437b29a7